src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -177,124 +196,169 @@ export interface SavedObjectsSetupDeps {
       savedObjectMigrations,
       savedObjectValidations,
     } = setupDeps.legacyPlugins.uiExports;
+    this.mappings = savedObjectMappings;
+    this.migrations = savedObjectMigrations;
+    this.schemas = savedObjectsSchemasDefinition;
+    this.validations = savedObjectValidations;
 
-    const savedObjectSchemas = new SavedObjectsSchema(savedObjectsSchemasDefinition);
+    return {
+      setClientFactoryProvider: provider => {
+        if (this.clientFactoryProvider) {
+          throw new Error(
+            'custom client factory is already set, unable to replace the current one'
+          );
+        }
+        this.clientFactoryProvider = provider;
+      },
+      addClientWrapper: (priority, id, factory) => {
+        this.clientFactoryWrappers.push({
+          priority,
+          id,
+          factory,
+        });
+      },
+    };
+  }
+
+  public async start(
+    core: SavedObjectsStartDeps,
+    migrationsRetryDelay?: number
+  ): Promise<InternalSavedObjectsServiceStart> {
+    if (!this.setupDeps) {
+      throw new Error('#setup() needs to be run first');
+    }
+
+    this.logger.debug('Starting SavedObjects service');
 
     const kibanaConfig = await this.coreContext.configService
       .atPath<KibanaConfigType>('kibana')
       .pipe(first())
       .toPromise();
-
     const savedObjectsConfig = await this.coreContext.configService
       .atPath<SavedObjectsConfigType>('migrations')
       .pipe(first())
       .toPromise();
+    const adminClient = this.setupDeps!.elasticsearch.adminClient;
+    const migrator = this.createMigrator(kibanaConfig, savedObjectsConfig, migrationsRetryDelay);
 
-    const adminClient = setupDeps.elasticsearch.adminClient;
-
-    const migrator = (this.migrator = new KibanaMigrator({
-      savedObjectSchemas,
-      savedObjectMappings,
-      savedObjectMigrations,
-      savedObjectValidations,
-      logger: this.coreContext.logger.get('migrations'),
-      kibanaVersion: this.coreContext.env.packageInfo.version,
-      config: setupDeps.legacyPlugins.pluginExtendedConfig,
-      savedObjectsConfig,
-      kibanaConfig,
-      callCluster: migrationsRetryCallCluster(
-        adminClient.callAsInternalUser,
-        this.coreContext.logger.get('migrations'),
-        migrationsRetryDelay
-      ),
-    }));
+    /**
+     * Note: We want to ensure that migrations have completed before
+     * continuing with further Core start steps that might use SavedObjects
+     * such as running the legacy server, legacy plugins and allowing incoming
+     * HTTP requests.
+     *
+     * However, our build system optimize step and some tests depend on the
+     * HTTP server running without an Elasticsearch server being available.
+     * So, when the `migrations.skip` is true, we skip migrations altogether.
+     */
+    const cliArgs = this.coreContext.env.cliArgs;
+    const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
+    await migrator.runMigrations(skipMigrations);
```

Comment:
Optional: migrations could take some time, it would be good to add start/end logging around the call.

Reply:
Do you think we should use `info`, or is `debug` enough?

Reply:
`debug` should be enough